### ISSUE TITLE
[question] Header-only libraries and package_id

### ISSUE THREAD

**Issue by EdwinC-fr** [conan-io/conan-center-index#1968]
Currently, range-v3 recipe is buggy on Windows/MSVC because it does not set the appropriate C++ flags (/permissive- /experimental:preprocessor) required by the library.
While the fix looks easy, I'm concerned about 2 things before submitting the fix.

Two questions: 

1. ABI Compatibility: range-v3 is header-only but if these flags affect the ABI (not sure), then downstream might be affected. In a more general case, if compilation flags that affect the ABI, depend on the compiler, should the compiler be in "settings" even in the header-only cases ?
2. If not, how can we test on the CI to detect this ? Running the test_package on every platform even if header_only ?

**Comment by uilianries** [comment]
> flags (/permissive- /experimental:preprocessor) required by the library.

You can pass it by [cppinfo](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#cpp-info) attributes *cxxflags* and *cflags*. It won't affect the package ID. As it requires recipe interference, you can use the env var *CXXFLAGS* instead, which will obtain the same result.

As it's header-only the package ID should the same for any package, because we don't have a different header in package folder.

**Comment by SSE4** [comment]
header-only packages will always have the same package id by design. this is true that certain libraries might be incompatible with certain compiler flags, but it's probably way too hard to detect - right now we don't model all possible compiler flags as settings.
recipe forcing ABI affecting compiler flags is definitely no-go - such things usually belong to the profile, you usually don't want a dependency suddenly changing these compiler flags for you.

**Comment by EdwinC-fr** [comment]
I get your point. But to make things a bit clearer to me, can we agree that:

* `/permissive-` and `/experimental:preprocessor` have the same consequence (in terms of ABI) as setting `/stdc++=..` . From the exact same code, enabling/disabling these flags may cause 1) some code to not compile 2) some code to compile into two different binaries (two phases name lookup affected...).

It makes sense that having a "recipe forcing ABI affecting compiler flags is definitely no-go" as many compiler flags may affect ABI and it is impossible to know them (just a macro definition may cause such a difference).

The way Conan handles the C++standard currently is with a profile `compiler.cppstd` and  `tools.check_min_cppstd`. Following the same principle,  the consumer thus has to define himself
CXXFLAGS="/permissive- /experimental:preprocessor" (this is a burden but I do not have see any better solution).
In the meantime, we should check that these flags are present and issue an error otherwise.
Do we have any facilities to check that a compiler flag is present a configure time  (here "/permissive-" and "/experimental:preprocessor" like for the standard version.

**Comment by jgsogo** [comment]
>  should the compiler be in "settings" even in the header-only cases ?

Yes, it is totally fine to add the `compiler` to the `settings` of a header-only package.

> 2. If not, how can we test on the CI to detect this ? Running the test_package on every platform even if header_only ?

Yes as well. This is something I really think we need to do in Conan Center. We need to run the `test_package` with every profile, otherwise we cannot know that the header-only is properly packaged. It will take us some time because there are other priorities

----

About those flags, if they affect the ABI they cannot be propagated from a recipe, I agree with @EdwinC-fr , they belong to settings and profiles. 

I don't know if the problem here is CCI running only one profile for these header-only libraries and failing because the chosen compiler is not modern enough... and then you need to add those flags so it compiles and runs the test_package and CCI generates the package.

Is this somehow the motivation for this issue? Does every compiler (even latest GCC/Clang) require a `/experimental` flag to use range-v3 headers?

**Comment by EdwinC-fr** [comment]
> Is this somehow the motivation for this issue? Does every compiler (even latest GCC/Clang) require a `/experimental` flag to use > range-v3 headers?

https://github.com/ericniebler/range-v3/blob/4d6a463bca51bc316f9b565edd94e82388206093/CMakeLists.txt#L33

The latest release 0.10, needs MSVC with
* /std:c++17 /permissive-
* a fully conforming preprocessor, so /experimental:preprocessor

This is not the case with others compilers because the code has workarounds if a specific feature is not supported by the compiler.

IMO, yes, we  have two problems:
* the CCI running only one profile for these header-only libraries. It looks ok to have 1 packager, but we should have more profiles for the test_package.
* In the MSVC case, we should have a way to test that "/permissive- /experimental:preprocessor" are there in the compile flags
  (in order to have an error in a early stage). Or something similar to the cmake "try_compile" or "compile_features" to see if the compiler is compatible ?

As a final note, this not a severe issue in the case of range-v3 as these flags wont be required in the future releases:
https://github.com/ericniebler/range-v3/commit/cce62b1bdf5b004628fa5ecf34d50d44637cd317

**Comment by jgsogo** [comment]
> As a final note, this not a severe issue in the case of range-v3 as these flags wont be required in the future releases:
> [ericniebler/range-v3@cce62b1](https://github.com/ericniebler/range-v3/commit/cce62b1bdf5b004628fa5ecf34d50d44637cd317)

I love reading this. This would be the way to go, we cannot use a different profile just for a recipe that requires some flags (CI is not prepared, and probably it would open some Pandora's box). 

About testing header-only libraries with more profiles, it is in the backlog, it will get to production eventually.

We will also run more profiles, we will add all stable C++ standards to every compiler-version, recipes can decide in the `configure()` method if the given compiler-version-cppstd supports some feature (yes, sometimes it is not enough, but it will improve current scenario).
